Paper: Empowering Learner - Teaching Reproducible Research with Open-Source Tools

khider opened the pull request:
If you are creating this PR in order to submit a draft of your paper, please name your PR with `Paper: <title>`. An editor will then add a `draft` label; this will trigger GitHub Actions to run automated checks on your paper and build a preview. You may then work to resolve failed checks and ensure the preview build looks correct. If you have any questions, please tag the proceedings team in a comment on your PR with `@scipy-conference/2025-proceedings`.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#preview-your-paper) for more information.

Hi, I am Sanhita Joshi, @sanhitamj  . I will serve as the editor for this submission. Reach out to me if any assistance needed.

khider committed:
Update myst.yml

khider committed:
paper

khider committed:
Merge branch '2025' into 2025

CommonClimate committed:
Edits to main.md

Smit-create commented:
Hi @khider, thanks for submitting the paper. I am Smit, and I will be reviewing your paper. I gave the paper a first look, and it looks fantastic to read through some of the modules mentioned in it.

Smit-create commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+
+Sharing research data, software, and workflow enhances reproducibility, collaboration, and the directions of future research [@doi:10.1002/2015EA000136] and is fundamental to building a FAIR open science ecosystem. The desire to share and reuse scientific data and software has grown in recent years as funders [e.g. @doi:10.1108/TG-03-2014-0008; @vandereynden2016; @Zuiderwijk2014] and publishers [ ...
+
+Facilitating Reproducible Open Geoscience (FROGS) is an U.S. National Science Foundation (NSF) funded initiative designed to speed up the integration of open source tools into the research and publication workflows of geoscientists. FROGS has offered three seven-week long courses on geoscience analysis in Python and R and FAIR publishing principles. The courses emphasized science publishing as an ...
```
```suggestion
Facilitating Reproducible Open Geoscience (FROGS) is a U.S. National Science Foundation (NSF) funded initiative designed to speed up the integration of open source tools into the research and publication workflows of geoscientists. FROGS has offered three seven-week long courses on geoscience analysis in Python and R and FAIR publishing principles. The courses emphasized science publishing as an integral part of a research product. Each course included a synchronous workshop that introduced overarching concepts, followed by asynchronous, detailed online materials combined with biweekly office hours designed to help participants integrate the resources into their practice. The asynchronous component was facilitated by the [LeapFROGS](https://linked.earth/LeapFROGS/) learning platform [@LeapFROGS], which provided summaries of the key concepts, curated additional resources from the open science community, and offered challenges for learners to assess their comprehension.
```

khider replied:
Changed

ameyxd commented:
Inviting reviewer: @kurtmckee

khider committed:
Update main.md

khider committed:
Update myst.yml

kurtmckee commented on papers/scipy-2025-Deborah_Khider/mybib.bib:
```diff
@@ -0,0 +1,293 @@
+@article{brandl2021sphinx,
+  title={Sphinx documentation},
+  author={Brandl, Georg},
+  journal={URL http://sphinx-doc. org/sphinx. pdf},
```
This citation doesn't look right. I'm not an expert on BibTex, and I don't have specific knowledge how to cite this documentation page, but two things jump out:

1. The URL shouldn't have spaces.
2. Without spaces, the URL isn't valid.

This is the URL that leads to a PDF file:

https://www.sphinx-doc.org/_/downloads/en/master/pdf/

But it may be worthwhile to link to the HTML documentation directly:

https://www.sphinx-doc.org/en/master/

khider replied:
Thank you for catching all the BibTeX error. Fixing these as I go. This one will be taken care of in the next commit (which may not be today but using comments to keep track).

kurtmckee commented on papers/scipy-2025-Deborah_Khider/mybib.bib:
```diff
@@ -0,0 +1,293 @@
+  author    = { {P}roject {J}upyter and {M}atthias {B}ussonnier and {J}essica {F}orde and {J}eremy {F}reeman and {B}rian {G}ranger and {T}im {H}ead and {C}hris {H}oldgraf and {K}yle {K}elley and {G}ladys {N}alvarte and {A}ndrew {O}sheroff and {M} {P}acer and {Y}uvi {P}anda and {F}ernando {P}erez and {B}enjamin {R}agan-{K}elley and {C}arol {W}illing },
+  title     = { {B}inder 2.0 - {R}eproducible, interactive, sharable environments for science at scale },
+  booktitle = { {P}roceedings of the 17th {P}ython in {S}cience {C}onference },
+  pages     = { 113 - 120 },
```
In other places, the `pages` field uses a double-hyphenated syntax, like `{113--120}`.

Should this match?

khider replied:
Added

kurtmckee commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+# Ensure that this title is the same as the one in `myst.yml`
+title: Empowering Learners - Teaching Reproducible Research with Open-Source Tools
+abstract: |
+  Reproducibility and open science are increasingly recognized as essential components of rigorous geoscientific research. However, gaps in training often hinder early-career scientists from fully adopting best practices in computational workflows and FAIR (Findable, Accessible, Interoperable, Reusable) publishing. The Facilitating Reproducible Open Geoscience (FROGS) initiative addresses this ne ...
```
It seems like verb tense changes a few times across sentences.

> The FROGS initiative _addresses_ this need...
> Across three sequential workshops, participants _engaged_ in hands-on learning activities...
> Survey results _demonstrate_ high participant satisfaction...


I think the middle sentence, "participants engaged", may benefit from a bridging sentence. Reading the abstract, I wasn't clear if the FROGS initiative was ongoing, or was a one-time set of workshops that is now over, or what.

If it's an ongoing initiative, it may be helpful to note that this survey was for a particular run of the workshops.

khider replied:
Rewrote the middle as:


> The Facilitating Reproducible Open Geoscience (FROGS) initiative addresses this need by offering an ongoing series of integrated courses and self-paced online modules designed to build capacity in Python and R programming, time series analysis, and open science publishing within the geosciences. To date, three sequential courses have been completed, and this paper reports on these initial offerings. During the courses, participants engaged in hands-on learning activities combining synchronous instruction with asynchronous exercises, supported by the LeapFROGS platform.

kurtmckee commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+
+Sharing research data, software, and workflow enhances reproducibility, collaboration, and the directions of future research [@doi:10.1002/2015EA000136] and is fundamental to building a FAIR open science ecosystem. The desire to share and reuse scientific data and software has grown in recent years as funders [e.g. @doi:10.1108/TG-03-2014-0008; @vandereynden2016; @Zuiderwijk2014] and publishers [ ...
+
+Facilitating Reproducible Open Geoscience (FROGS) is a U.S. National Science Foundation (NSF) funded initiative designed to speed up the integration of open source tools into the research and publication workflows of geoscientists. FROGS has offered three seven-week long courses on geoscience analysis in Python and R and FAIR publishing principles. The courses emphasized science publishing as an  ...
```
> Each course included a synchronous workshop

In the abstract, the word "workshop" is used to refer to the seven-week-long blocks. Here, it seems "course" is used, and "workshop" appears to refer to the synchronous portions. If I'm understanding that correctly, could these terms be aligned between the abstract and the main body of text?

khider replied:
Rephrase as courses in the abstract (see above comment).

kurtmckee commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+
+## Introduction and Motivation
+
+Sharing research data, software, and workflow enhances reproducibility, collaboration, and the directions of future research [@doi:10.1002/2015EA000136] and is fundamental to building a FAIR open science ecosystem. The desire to share and reuse scientific data and software has grown in recent years as funders [e.g. @doi:10.1108/TG-03-2014-0008; @vandereynden2016; @Zuiderwijk2014] and publishers [ ...
```
> Sharing research data, software, and workflow

I was anticipating "workflows" here: "data, software, and workflows". What do you think?

----

> This training needs to take place across science practice and publishing.

The phrase "science practice" caught my eye. Elsewhere in the paper, I've found that the phrase "scientific practice" is used. Could this be phrased as "scientific practice"? I think it will aid readability.

----

Separately, I generally recommend breaking each sentence onto its own line in Markdown files. This is helpful in the context of a GitHub review because it makes it easier to leave a comment on a specific line. (With the sentences combined together into a long paragraph, it's harder for reviewers to point out a specific line or phrasing while providing feedback, and as you can see in the code change suggestion below, the line is so long that GitHub doesn't highlight the changes that I've suggested.)

----

```suggestion
Sharing research data, software, and workflows enhances reproducibility, collaboration, and the directions of future research [@doi:10.1002/2015EA000136] and is fundamental to building a FAIR open science ecosystem. The desire to share and reuse scientific data and software has grown in recent years as funders [e.g. @doi:10.1108/TG-03-2014-0008; @vandereynden2016; @Zuiderwijk2014] and publishers [e.g. @AGUSoftwareGuidelines; @AGUWorkflowGuidelines; @doi:10.1038/s41592-019-0350-x] have introduced open science policies emphasizing reproducibility, and as scientists increasingly recognize the benefits of open science [@doi:10.1038/s41559-017-0160; @doi:10.7554/elife.16800]. Consequently, the last decade has seen a proliferation of frameworks to promote open source resources, collaboration among scientists, and sharing of scientific research products. One critical element to the success and long-term sustainability of these resources is the training of scientists, especially in their early-career, in their use. This training needs to take place across scientific practice and publishing. 
```

kurtmckee commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+Example of training model on the LeapFROGS platform. (a) An exercise on numbers in Python with links to the Trinket Python course, and (b) an tutorial on Pandas using materials developed by Project Pythia. LeapFROGS tests understanding of these concepts by asking to fill in the blank in a code cell than can be executed on myBinder so learners can run their code to check against the solution. Lear ...
+:::
+
+The second module, *The scientific Python Stack*, introduces libraries common to scientific Python such as Jupyter [@jupyter], NumPy [@numpy], pandas [@pandas1; @pandas2], Matplotlib [@matplotlib], Cartopy [@Cartopy], Seaborn [@Seaborn], statsmodels [@statsmodels], scikit-learn [@sklearn1; @sklearn2], PyTorch [@pytorch],and xarray [@xarray]. This module adopts the same approach as the *Introducti ...
```
Suggesting "The Scientific Python Stack" (capitalized) and adding a space between a comma and the words "and xarray"

```suggestion
The second module, *The Scientific Python Stack*, introduces libraries common to scientific Python such as Jupyter [@jupyter], NumPy [@numpy], pandas [@pandas1; @pandas2], Matplotlib [@matplotlib], Cartopy [@Cartopy], Seaborn [@Seaborn], statsmodels [@statsmodels], scikit-learn [@sklearn1; @sklearn2], PyTorch [@pytorch], and xarray [@xarray]. This module adopts the same approach as the *Introduction to Python* module, featuring lessons from Project Pythia Foundations [@PythiaFoundations], alongside tutorials from library developers and coding exercises tailored to reinforce learning within a geoscience context (@fig:python). 
```

kurtmckee commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+
+The fourth module, *The Scientific Paper of The Future*, introduces concepts in FAIR scientific publishing. As its name indicates, this module presents an updated version [@KhiderGPF] of the Geoscience Paper of the Future (GPF) [@doi:10.1002/2015EA000136]. The GPF represented one of the earliest community efforts to fully document, share, and cite all research products, including data, software,  ...
+
+The fifth module, *Using GitHub for your research*, focuses on GitHub as a platform to collaborate, share scientific workflow and software, and support project management. The module covers the basics of repositories, forks, pull requests, setting up actions for automation, and linking a GitHub repository to Zenodo to obtain a persistent identifier. Lessons from this chapter were summarized from  ...
```
Prior module names are title-cased; it may be worthwhile to follow that for this and the last one for consistency.

```suggestion
The fifth module, *Using GitHub for Your Research*, focuses on GitHub as a platform to collaborate, share scientific workflow and software, and support project management. The module covers the basics of repositories, forks, pull requests, setting up actions for automation, and linking a GitHub repository to Zenodo to obtain a persistent identifier. Lessons from this chapter were summarized from various sources including the GitHub documentation, blogs from Medium and other sources, and Project Pythia Foundations [@PythiaFoundations].
```

kurtmckee commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+
+The sixth module, *Sharing Reproducible Workflows*, provides information about using [Docker](https://www.docker.com) and [myBinder](https://mybinder.org) to share reproducible science results. 
+
+Finally, the last module, *Packaging your software for sharing* walks through instructions and tutorials on creating Python packages, publishing it on software registries (e.g., The [Python Package Index repository](https://pypi.org), PyPi), on creating documentation using Sphinx [@brandl2021sphinx] and publishing it on [readthedocs.org](https://readthedocs.org), on unit testing, on setting up Gi ...
```
"PyPI" uses the capitalization I've used in this sentence, and it's simply called the "Python Package Index" -- no need for the word "repository" at the end.

```suggestion
Finally, the last module, *Packaging Your Software for Sharing* walks through instructions and tutorials on creating Python packages, publishing it on software registries (e.g., The [Python Package Index](https://pypi.org), PyPI), on creating documentation using Sphinx [@brandl2021sphinx] and publishing it on [readthedocs.org](https://readthedocs.org), on unit testing, on setting up GitHub actions for continuous integration, and on publishing on PyPI. Most of these tutorials were based on the materials developed by the [PyOpenSci](https://www.pyopensci.org) community. 
```

I wanted to double-check the reference to GitHub actions here. Module 5 states that participants are introduced to GitHub actions for automation; this is a more specific phrasing, but were GitHub actions covered in both modules 5 and 7?

khider replied:
Modules 5, 6, and 7. 

- Module 5 covered them at high level (what they are, why it's done this way and the basic anatomy of a workflow file 
- Module 6 specifically talks about setting up GitHub actions to automatically create and upload containers to either DockerHub or Quay.io
- Module 7 talks about using GitHub actions for CI specifically. Rephrased in the various module description in the paper. 

Paragraph describing module 5 has been rewritten as follows:

> The fifth module, *Using GitHub for Your Research*, focuses on GitHub as a platform to collaborate, share scientific workflow and software, and support project management. The module covers the basics of repositories, forks, pull requests, understanding and setting up actions for automation, and linking a GitHub repository to Zenodo to obtain a persistent identifier. This module provides a high-level overview of GitHub Actions, including their purpose and the basic structure of a workflow file, while more detailed applications—such as containerization and continuous integration—are addressed in the final two modules.Lessons from this chapter were summarized from various sources including the GitHub documentation, blogs from Medium and other sources, and Project Pythia Foundations [@PythiaFoundations].

kurtmckee commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+
+#### Open Geoscience Hackathon
+
+The Open Geoscience Hackathon targeted researchers interested in open-source software development and contributions. The workshop focused on practical skills including opening pull requests, software packaging, unit testing, and continuous integration (module 7 on LeapFROGS). Held in person at USC ISI in May 2025, the hackathon gathered 17 participants. Morning sessions were reserved for high-lev ...
```
"GitHub"

```suggestion
The Open Geoscience Hackathon targeted researchers interested in open-source software development and contributions. The workshop focused on practical skills including opening pull requests, software packaging, unit testing, and continuous integration (module 7 on LeapFROGS). Held in person at USC ISI in May 2025, the hackathon gathered 17 participants. Morning sessions were reserved for high-level lectures and guided homework on the PyCatSim [@PyCatSim] toy package. This Python package, which simulates cat ownership, was specifically designed for this hackathon. The functions were straightforward to implement and served to illustrate object-oriented programming principles, methods for integrating data into packages, and error handling, all while demonstrating the process of contributing to an open-source project. In the afternoons, participants worked on their own package. By the end of the workshop, participants were able to share the foundation of their package on GitHub.
```

kurtmckee commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+(result-section)=
+## Contributions to open science
+
+FROGS has contributed to open science in two main ways. First, by training the next generation of scientists in Python programming, data science libraries, timeseries analysis, data science in the geoscience community, and open science sharing and publishing practices. In turn, these scientists produced exemplars of reproducible studies in their respective community. 
```
The "First, by training..." sentence is a sentence fragment. It's fine in a casual context, but I wanted to call it out in case this needs to be rephrased for publication.

----

Should the end of the third sentence be "in their respective communities"?

khider replied:
Rewritten as:

> FROGS has contributed to open science in two main ways. First, it has trained the next generation of scientists in Python programming, data science libraries, time series analysis, geoscience data science, and open science sharing and publishing practices. In turn, these scientists have produced exemplars of reproducible studies within their respective communities.

kurtmckee reviewed (changes requested):
Hello! The FROGS initiative sounds great, and I'm particularly impressed with the hands-on efforts that appear to have benefited several open source projects!

I've left feedback about some of the content. Please let me know if you have questions, or if I misread something and the feedback isn't relevant.

khider committed:
Update papers/scipy-2025-Deborah_Khider/main.md

Agreed and committed the suggestion.

Co-authored-by: Kurt McKee <contactme@kurtmckee.org>

khider committed:
edits in reponse to reviewer comments

khider committed:
Update papers/scipy-2025-Deborah_Khider/main.md

Capitalize Scientific Python Stack

Co-authored-by: Kurt McKee <contactme@kurtmckee.org>

khider committed:
Update papers/scipy-2025-Deborah_Khider/main.md

Capitalize 'Your Research'

Co-authored-by: Kurt McKee <contactme@kurtmckee.org>

khider committed:
Update papers/scipy-2025-Deborah_Khider/main.md

Committing the changes to typos. Need to further clarify the GitHub actions parts.

Co-authored-by: Kurt McKee <contactme@kurtmckee.org>

khider committed:
updates to module description

khider committed:
Update papers/scipy-2025-Deborah_Khider/main.md

Capitalize GitHub

Co-authored-by: Kurt McKee <contactme@kurtmckee.org>

kurtmckee commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+
+The fourth module, *The Scientific Paper of The Future*, introduces concepts in FAIR scientific publishing. As its name indicates, this module presents an updated version [@KhiderGPF] of the Geoscience Paper of the Future (GPF) [@doi:10.1002/2015EA000136]. The GPF represented one of the earliest community efforts to fully document, share, and cite all research products, including data, software,  ...
+
+The fifth module, *Using GitHub for Your Research*, focuses on GitHub as a platform to collaborate, share scientific workflow and software, and support project management. The module covers the basics of repositories, forks, pull requests, understanding and setting up actions for automation, and linking a GitHub repository to Zenodo to obtain a persistent identifier. This module provides a high-l ...
```
One last thing that I spotted in the latest set of changes:

```
final two modules.Lessons
                ^^^
```

This suggestion adds the missing space.

```suggestion
The fifth module, *Using GitHub for Your Research*, focuses on GitHub as a platform to collaborate, share scientific workflow and software, and support project management. The module covers the basics of repositories, forks, pull requests, understanding and setting up actions for automation, and linking a GitHub repository to Zenodo to obtain a persistent identifier. This module provides a high-level overview of GitHub Actions, including their purpose and the basic structure of a workflow file, while more detailed applications—such as containerization and continuous integration—are addressed in the final two modules. Lessons from this chapter were summarized from various sources including the GitHub documentation, blogs from Medium and other sources, and Project Pythia Foundations [@PythiaFoundations].
```

khider committed:
Update papers/scipy-2025-Deborah_Khider/main.md

Co-authored-by: Kurt McKee <contactme@kurtmckee.org>

kurtmckee reviewed (approved):
Looks good!

ameyxd commented:
@sanhitamj will serve as editor for this paper.

Smit-create reviewed (approved):
Thanks for the submission. Looks good to me.

sanhitamj commented on papers/scipy-2025-Deborah_Khider/mybib.bib:
```diff
@@ -0,0 +1,293 @@
+
+@misc{trinket,
+  title = {Getting started with Python},
+  author ={{trinket}},
```
T in upper case?

khider replied:
Added upper case.

sanhitamj commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+Example of training model on the LeapFROGS platform. (a) An exercise on numbers in Python with links to the Trinket Python course, and (b) an tutorial on Pandas using materials developed by Project Pythia. LeapFROGS tests understanding of these concepts by asking to fill in the blank in a code cell than can be executed on myBinder so learners can run their code to check against the solution. Lear ...
+:::
+
+The second module, *The Scientific Python Stack*, introduces libraries common to scientific Python such as Jupyter [@jupyter], NumPy [@numpy], pandas [@pandas1; @pandas2], Matplotlib [@matplotlib], Cartopy [@Cartopy], Seaborn [@Seaborn], statsmodels [@statsmodels], scikit-learn [@sklearn1; @sklearn2], PyTorch [@pytorch], and xarray [@xarray]. This module adopts the same approach as the *Introduct ...
```
`Xarray` is how they fashion it.

khider replied:
changed

sanhitamj commented on papers/scipy-2025-Deborah_Khider/main.md:
```diff
@@ -0,0 +1,139 @@
+(conclusion-section)=
+## Conclusions
+
+In total, the workshops engaged 56 participants spanning over a dozen geoscience subfields and multiple career stages. Participants produced 8 reproducible notebooks and contributed 9 open-source repositories for software development, underscoring tangible skill acquisition and active engagement with open science practices. These contributions not only exemplify the immediate outputs of the train ...
```
> ... contributed 9 open-source repositories ...
contributed `to` 9 (existing) open source repositories or contributed 9 (new) open-source repositories?

khider replied:
changed.

khider committed:
Typo fixes
